Add action to deploy Parse Cloud Code

teddynewell opened the pull request:
Adds an action to deploy Parse cloud code with optional application, release notes, and path to /Parse directory.

This action can be used for projects that use [Parse](https://parse.com) for backend infrastructure.

Commit:
Added parse_deploy action to deploy Parse Cloud Code

- Action accepts target application parameter and custom path to Parse
project directory

Commit:
Added support for Parse release notes in parse_deploy action

Commit:
Fixes for Ruby style guidelines

Commit:
Merge remote-tracking branch 'fastlane/master'

Commit:
Removed trailing `.` from action description to meet format requirements

Commit:
Added instructional comment to custom_action_template::description

- Noted that custom_action_template::description should not end with
trailing punctuation
- Fixed a typo in the custom_action_template::details instructional
comment

KrauseFx commented on lib/assets/custom_action_template.rb:
```diff
@@ -25,12 +25,14 @@ def self.run(params)
       #####################################################
 
       def self.description
+        # Give a short description (<= 80 characters) for this action
```
:+1:

KrauseFx commented on lib/fastlane/actions/parse_deploy.rb:
```diff
@@ -0,0 +1,103 @@
+      #####################################################
+
+      def self.description
+        "Deploy Parse Cloud code"
```
This should make clear it's about server code, and has nothing to do with push notifications maybe?

teddynewell replied:
Clarified description and details in 484efc7

KrauseFx commented on lib/fastlane/actions/parse_deploy.rb:
```diff
@@ -0,0 +1,103 @@
+        parse_directory = params[:parse_directory]
+        release_notes = params[:release_notes]
+
+        if !parse_application || parse_application.empty?
```
Replace with 

```ruby
parse_application.to_s.length == 0
```

teddynewell replied:
Replaced in 0faab33

KrauseFx commented on lib/fastlane/actions/parse_deploy.rb:
```diff
@@ -0,0 +1,103 @@
+                                       description: "Target Application for ParseDeployAction", # a short description of this parameter
+                                       is_string: false, # true: verifies the input is a string, false: every kind of value
+                                       optional: true,
+                                       default_value: ""), # the default value if the user didn't provide one
```
You can just get rid of this line and the default_value will be `nil`

KrauseFx replied:
Oh I see you actually want an empty string, never mind then

KrauseFx commented:
This is great, thank you for your contribution - I added some comments before it's ready to be merged.

Out of curiosity, how do you use it in your project? You deploy your iOS app and you back-end at the same time using `fastlane`?

KrauseFx commented:
Please also update `Actions.md` with a short description on how to use this integration :+1:

Commit:
Clarifies action description and details

- Specifies the action is for deploying server-side code to Parse Cloud

Commit:
Simplifies parameter check to identify the default empty-string case

Commit:
Added summary of action in docs/Actions.md

teddynewell commented:
Updated `Actions.md` in aafd394

teddynewell commented:
We just started using `fastlane` to manage our synchronized Development -> Staging -> Production deployment.  Before pushing nightly and TestFlight builds, we make sure any back-end changes are deployed to the appropriate environment.  It's working well for our bootstrap project.

Very happy with the tool suite so far – thanks for making it :smile:

KrauseFx commented on docs/Actions.md:
````diff
@@ -774,6 +774,25 @@ appetize(
+parse_deploy
+```
+
+You can pass all options listed in `parse_deploy --help` in `fastlane`:
````
Please remove this part, as it's described on the top of the document :+1:

teddynewell replied:
Removed in e2732a1

KrauseFx commented on lib/fastlane/actions/parse_deploy.rb:
```diff
@@ -0,0 +1,103 @@
+        exit_code = 1
+        if File.exist?(parse_path)
+          Dir.chdir(parse_path) do
+            command = "Parse deploy"
```
Shouldn't `Parse` be lower case?

KrauseFx replied:
What if this tool is not installed? Maybe we should add something like
```ruby
raise "Please install `parse` using `...`" unless `parse --help`.include?("something here")
```

teddynewell replied:
Good thought – added in cdad8c6

KrauseFx commented on lib/fastlane/actions/parse_deploy.rb:
```diff
@@ -0,0 +1,103 @@
+            if release_notes && !release_notes.empty?
+              command << " --description=\"#{release_notes}\""
+            end
+            Helper.log.info "Deploy command: #{command}".blue
```
Please change to `.yellow`, which is the default color for commands

teddynewell replied:
Changed in cdad8c6

KrauseFx commented:
That's so great, thank you so much for this amazing pull request. I added a few more comments before it's ready to be merged, sorry for being so picky :wink:

Commit:
Removed redundant information in Actions.md

Commit:
Merge remote-tracking branch 'fastlane/master'

Commit:
Added action failure case and improved log output

- Changed command output to yellow and to only log the command
- Fixed a capitalization error
- Added check for the `parse` tool with an exception providing
instructions for parse installation if needed

Commit:
Replaced `unless !` with `if` for best practices

teddynewell commented:
Responded to your helpful comments – I'm not too familiar with Ruby, so I appreciate the suggestions.

KrauseFx commented:
Ah, thanks @teddynewell. Somehow this ended up in a merge conflict. Sorry about the the trouble :wink:

Commit:
Merge remote-tracking branch 'fastlane/master'

Fix for merge conflict from multiple additions at same location in
docs/Actions.md
# Conflicts:
#	docs/Actions.md

teddynewell commented:
Updated to fix the merge conflict :)

KrauseFx commented:
@teddynewell thanks for updating the PR. One last thing before it's ready to be merged: could you please squash your commits to one? :+1:

Commit:
Merge remote-tracking branch 'fastlane/master'

KrauseFx commented:
Hey @teddynewell, just wanted to check if you had time to squash the commits? Would love to have this included in fastlane :rocket:

teddynewell commented:
Thanks for the followup – haven’t had time yet, as I spaced out on my clone and made changes on master.  I’ll figure it out and update the PR soon :)
> On Jan 21, 2016, at 4:58 PM, Felix Krause <notifications@github.com> wrote:
> 
> Hey @teddynewell <https://github.com/teddynewell>, just wanted to check if you had time to squash the commits? Would love to have this included in fastlane 
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/fastlane/fastlane/pull/967#issuecomment-173766118>.
>

ohwutup commented:
hey @teddynewell! i am going to close this PR for now as it's been inactive for some time, but please feel free to open another PR when you get the time to do the final cleanup and squash!! thanks again!

KrauseFx commented:
Thanks for updating this issue @ohwutup. Correct me if I'm wrong, but Parse will shutdown soon, so there is no real value in working on this, right?

ohwutup commented:
oh, you're right. not sure why i didn't think of that, although, they did open source their platform right?

teddynewell commented:
Thanks for closing this @ohwutup – I de-prioritized the last cleanup to the PR since the Parse shutdown until I can investigate if the open-source version has the same command line tool.  If it does, I'll create a new PR with the final squashing.
